CVLByExamples/ConstantProductPoolExample/certora/spec/ConstantProductPool.spec

```diff
@@ -0,0 +1,264 @@
+    
+    uint256 before = balanceOf(account);
+    calldataarg args;
+    f(e,args); /* check on all possible arguments */
+    uint256 after = balanceOf(account);
+    /* logic implication : true when: (a) the left hand side is false or (b) right hand side is true  */
+    assert after < before =>  (e.msg.sender == account  ||  to_mathint(allowance) >= (before-after))  ;
+}
+
+
+/*
+Property: For both token0 and token1 the balance of the system is at least as much as the reserves.
+
+This property is implemented as an invariant. 
+Invariants are a specification of a condition that should always be true once an operation is concluded.
+In addition, the invariant also checks that it holds right after the constructor of the code runs.
+
+This invariant also catches the bug in which there is a switch between the token and the recipient in burnSingle:
+        transfer( recipient, tokenOut, amountOut);
+
+Formula:
+    getReserve0() <= _token0.balanceOf(currentContract) &&
+    getReserve1() <= _token1.balanceOf(currentContract)
+*/
+
+invariant balanceGreaterThanReserve()
+    (getReserve0() <= _token0.balanceOf(currentContract))&&
+    (getReserve1() <= _token1.balanceOf(currentContract))
+    {
+        preserved with (env e){
+         setup(e);
+        }
+    }
+
+
+/*
+Property: Integrity of totalSupply with respect to the amount of reserves. 
+
+This is a high level property of the system - the ability to pay back liquidity providers.
+If there are any LP tokens (the totalSupply is greater than 0), then neither reserves0 nor reserves1 should ever become zero (otherwise the pool could not produce the underlying tokens).
+
+This invariant catches the original bug in Trident where the amount to receive is computed as a function of the balances and not the reserves.
+
+Formula:
+    (totalSupply() == 0 <=> getReserve0() == 0) &&
+    (totalSupply() == 0 <=> getReserve1() == 0)
+*/
+
+invariant integrityOfTotalSupply()
+    
+    (totalSupply() == 0 <=> getReserve0() == 0) &&
+    (totalSupply() == 0 <=> getReserve1() == 0)
+    {
+        preserved with (env e){
+            requireInvariant balanceGreaterThanReserve();
+            setup(e);
+        }
+    }
+
+
+/*
+Property: Monotonicity of mint.
+
+The more tokens a user transfers to the system the more LP tokens that user should receive. 
+This property is implemented as a relational property - it compares two different executions on the same state.
+
+This invariant catches a bug in mint where the LP tokens of the first depositor are not computed correctly and the less he transfers the more LP-tokens he receives. 
+
+Formula:
+    { x > y }
+        _token0.transfer(currentContract, x); mint(recipient);
+        ~ 
+        _token0.transfer(currentContract, y); mint(recipient);
+    { balanceOf(recipient) at 1  >=  balanceOf(recipient) at 2  }
+*/
+
+rule monotonicityOfMint(uint256 x, uint256 y, address recipient) {
+    env eT0;
```

Comment:
I would give these more meaningful names, and also move them closer to where they are used.